### ISSUE TITLE
Exempt requests should still count into flowcontrol so other requests get pushed out

### ISSUE THREAD

**Issue by grosser** [kubernetes/kubernetes#94907]
We saw our api servers falling over because excempt + regular requests put too much load onto them, it would be great if excempt requests would still count into flowcontrol so that regular requests get throttled and the overall max-requests-in-flight is honored. Because otherwise it's "infinite excempt requests + in-flight" which is hard to configure right.

/sig api-machinery

**Comment by yue9944882** [comment]
the system flow rules will regard all the traffic from `system:masters` user-group as exempt, which by default only covers the apiserver's loopback requests. can you verify if there're other unexpected requests count as exempt?

on the other way, i don't think currently intercepting exempt traffic to a custom priority-level is supported. every modification you made upon exempt flow rule will be reconciled back periodically b/c they're designed to be mandatory..

**Comment by lavalamp** [comment]
The closest approximation of the exact request above is to have in-flight exempt requests dynamically deduct the global concurrency value (the one set in the flag), but I'm not sure that's the right solution. The idea is that we want to pay for the requests up front, when the request which causes the self-requests is admitted.

We're currently thinking up a "width" concept to account for some requests being more expensive than others. I think the solution here has two parst:
1. Some way to figure out the "cause" of the exempt request.
2. Some way to adjust the "width" of source request. This could be manual, but automatic adjustment is preferred.

To define my terms a bit more, a "cause" for an exempt request X is some other request Y. In the course of processing request Y, the system had to issue request X against itself. This could be apiserver itself issued (e.g., get a secret token to do an authn check on an incoming request), or it could be issued by a webhook (look up something in order to respond). A special case is the DELETECOLLECTION verb which causes apiserver to issue many deletion requests to itself.

(This can be a DAG of requests where Z causes Y causes X. It cannot be a graph with cycles in it because we haven't invented time machines.)

The first take of this idea is to keep a running average, probably per-flow-schema, of how many requests are caused by requests matching that flow schema. So, we wish to find the root cause request for every exempt request, and update its count.

Then, when admitting requests from a flowschema, we'll use this count to estimate their "width".

This is fairly trivial to implement for the apiserver self-request case, but when webhooks and multiple apiservers are involved, we don't currently have a mechanism that lets one apiserver know what requests another apiserver is handling, so I recommend we solve that at a later time; I think it's not a big problem anyway, at least not yet.

So, anyway, that'll be a little longer term. To actually be useful for the question at hand: if you can figure out what requests are causing the exempt requests, you can put them in a priority level without many shares, and hopefully that will solve things--please let us know.

**Comment by grosser** [comment]
@yue9944882 / @lavalamp  this is about "loopback" requests (kube-controller-manager running on the same node and connecting via localhost)
the issue we saw was that traffic alone overloaded the api server and then it stopped doing etcd compaction and then the cluster fell over, so having the excempt request also count into the global rate-limit should have prevented that

the requests are caused by lots of new deployments starting at the same time, which I guess would be the point were we could throttle it ... but that's a bit clunky since deployments are not the same size

FYI we remediated the immediate issue by making etcd run an emergency compaction when api server fails to do so

**Comment by lavalamp** [comment]
Oh... You have controller manager using the insecure port?

I recommend stopping that as soon as possible, it's being removed soon (actually I thought it was already).

In general you don't want controller manager, scheduler, etc sending exempt requests-- protecting the system from them is a large part of this feature's purpose :)

It's specifically re-entrant webhook requests (i.e., 3 in this chain: user ---1---> apiserver ---2---> webhook ---3---> apiserver) that should be exempt.

**Comment by grosser** [comment]
We plan to switch that to the scure loopback (avoid going over the network
/ stay local)
... that's correct or is there some other guidance on that ?

**Comment by lavalamp** [comment]
Yeah, that's fine.

**Comment by MikeSpreitzer** [comment]
We will be able to handle recursive requests better after #94942 merges.

**Comment by lavalamp** [comment]
/reopen
/lifecycle frozen

I think it's probably still important that exempt requests reduce the available concurrency for non-exempt requests.

@wojtek-t in case you haven't seen this one.

**Comment by wojtek-t** [comment]
Thanks for cc-ing Daniel. I was aware of the problem, but not about this issue.

**Comment by MikeSpreitzer** [comment]
We just discussed this in the APF meeting and had a few thoughts.

Now that APF has borrowing between priority levels, the exempt requests could be measured and borrow from the non-exempt requests in the regular borrowing calculation.  That would throttle the non-exempt ones down in response to usage by exempt requests.  However, that alone does not impose a limit on the exempt requests.

Imposing a limit on the exempt requests is dangerous because requests critical for basic health and problem recovery are exempt and blocking them can create priority inversions.  If the apiservers cannot handle the load that is critical for basic health then the problem is the true overload, not any failure in regulation.  If the problem is not a true overload then perhaps the problem is not creating less privileged users for some operations.

**Comment by lavalamp** [comment]
I think limiting exempt requests is a non-goal, but it does seem good for exempt requests to crowd out non-exempt requests to the extent necessary (if exempt+non-exempt > concurrency limit) or possible (when exempt >= concurrency limit, you can't balance that out by dropping non-exempt requests). Borrowing an unlimited amount from other PLs seems like a fine way to model this.

**Comment by wojtek-t** [comment]
This was already addressed via appropriate borrowing in 1.28.

/cc @MikeSpreitzer @tkashem 

/close
